tests/test_tools.py

```diff
@@ -27,5 +32,114 @@ def test_ClientRedirectServer(self):
+    @mock.patch('oauth2client.tools.logging')
+    @mock.patch('oauth2client.tools.ClientRedirectServer')
+    @mock.patch('webbrowser.open')
+    def test_run_flow_webserver(
+            self, webbrowser_open_mock, server_ctor_mock, logging_mock):
+        server = mock.Mock()
+        flow = mock.Mock()
+        storage = mock.Mock()
+        credentials = mock.Mock()
+
+        server_ctor_mock.return_value = server
+        server.query_params = {'code': 'auth_code'}
+        flow.step1_get_authorize_url.return_value = 'http://example.com/auth'
+        flow.step2_exchange.return_value = credentials
+
+        # Successful exchange.
+        returned_credentials = tools.run_flow(flow, storage)
+
+        self.assertEqual(credentials, returned_credentials)
+        self.assertEqual(flow.redirect_uri, 'http://localhost:8080/')
+        flow.step2_exchange.assert_called_once_with('auth_code', http=None)
+        storage.put.assert_called_once_with(credentials)
+        credentials.set_store.assert_called_once_with(storage)
+        webbrowser_open_mock.assert_called_once_with(
+            'http://example.com/auth', autoraise=True, new=1)
+        self.assertTrue(server.handle_request.called)
+
+        # Exchange returned an error code.
+        with self.assertRaises(SystemExit):
+            server.query_params = {'error': 'any error'}
+            returned_credentials = tools.run_flow(flow, storage)
+
+        # No code found in response
+        with self.assertRaises(SystemExit):
+            server.query_params = {}
+            returned_credentials = tools.run_flow(flow, storage)
+
+    @mock.patch.object(sys, 'argv', [sys.executable,])
+    @mock.patch('oauth2client.tools.logging')
+    @mock.patch('oauth2client.tools.ClientRedirectServer')
+    @mock.patch('oauth2client.tools.input')
+    def test_run_flow_webserver_fallback(
+            self, input_mock, server_ctor_mock, logging_mock):
+        flow = mock.Mock()
+        storage = mock.Mock()
+        credentials = mock.Mock()
+
+        server_ctor_mock.side_effect = socket.error()
+        input_mock.return_value = 'auth_code'
+        flow.step1_get_authorize_url.return_value = 'http://example.com/auth'
+        flow.step2_exchange.return_value = credentials
+
+        # It should catch the socket error and proceed as if
+        # noauth_local_webserver was specified.
+        returned_credentials = tools.run_flow(flow, storage)
+
+        self.assertEqual(credentials, returned_credentials)
+        self.assertEqual(flow.redirect_uri, OOB_CALLBACK_URN)
+        flow.step2_exchange.assert_called_once_with('auth_code', http=None)
+        storage.put.assert_called_once_with(credentials)
+        credentials.set_store.assert_called_once_with(storage)
+        self.assertTrue(server_ctor_mock.called)
+
+
+class TestMessageIfMissing(unittest2.TestCase):
+    def test_message_if_missing(self):
+        self.assertIn('somefile.txt', tools.message_if_missing('somefile.txt'))
```

Comment:
samples that use `tools.run_flow`, I guess? I'll see if I can find actual users at another time.